flang/lib/Lower/ConvertExprToHLFIR.cpp

```diff
@@ -1425,9 +1425,14 @@ class HlfirBuilder {
   }
 
   hlfir::EntityWithAttributes gen(const Fortran::evaluate::ProcedureRef &expr) {
-    TODO(
-        getLoc(),
-        "lowering function references that return procedure pointers to HLFIR");
+    fir::FirOpBuilder &builder = getBuilder();
+    Fortran::evaluate::ProcedureDesignator proc{expr.proc()};
+    auto procTy{Fortran::lower::translateSignature(proc, getConverter())};
+    mlir::Type resType = fir::BoxProcType::get(builder.getContext(), procTy);
```

Comment:
I am not convinced that resType is correct here.

I would expect resType to be something like `procTy.getResult(0)``, I am wondering if the current code isn't creating a `fir.boxproc<(ArgsX) -> (fir.boxproc((R) -> (ArgsY, ...)))>` instead of `fir.boxproc((R) -> (ArgsY, ...))`.

I think this makes little functional difference currently given resType is not used to generate IR (just to check if it is a fir::BoxProcType, which is true in both case), but could you check that recType is the expected `fir.boxproc((R) -> (ArgsY, ...))` here to avoid future bugs?

Reply:
@jeanPerier Thanks for the review! Sorry for the delayed response. 
The `resType` as is here is `!fir.boxproc<(!fir.ref<i32>) -> !fir.boxproc<(!fir.ref<i32>) -> f32>>` for a function that returns a procedure pointer that has an interface of `real foo(integer)`
The test case I used for generating the code is
```
  interface
    real function func(x)
      integer :: x
    end
  end interface
  real :: res
  procedure(func), pointer :: arg0
  arg0=>reffunc(5)
contains
  function reffunc(arg1) result(pp)
    integer :: arg1
    procedure(func), pointer :: pp
    pp => func
  end
end
```

The FIR looks like:
```
...
%3:2 = hlfir.declare %0 {fortran_attrs = #fir.var_attrs<pointer>, uniq_name = "_QFEarg0"} : (!fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>) -> (!fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>, !fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>)
...
%7 = fir.call @_QFPreffunc(%6#1) fastmath<contract> : (!fir.ref<i32>) -> !fir.boxproc<(!fir.ref<i32>) -> f32> fir.store %7 to %3#0 : !fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>
```

Reply:
Actually, after further investigation, you are right. `procTy.getResult(0)` indeed gives the correct return type of the function that returns the procedure pointer.